alf/optimizers/optimizers.py

```diff
@@ -192,6 +201,74 @@ def __init__(self,
             self.name = NewClsName + "_" + str(NewCls.counter)
             NewCls.counter += 1
 
+    @common.add_method(NewCls)
+    def _generate_capacity_mask(self, capacity_ratio: float):
+        """generate capacity mask according to the specified capacity_ratio
+
+        Args:
+            capacity_ratio: the capacity ration specifying the ratio of learnable parameters
+                to the number of all parameters
+        Returns:
+            states: the states containing:
+                - rng_state: random number generator state
+                - mask: the mask specifying the non-learnable (when mask value is True),
+                learnable (when mask value is False) parameters
+                - old_param: the value of the parameters before gradient step
+        """
+        states = {}
+        if capacity_ratio < 1:
+            # To achieve this, we assign a random number for each element of
+            # the parameter. An element is turned on if its assigned random number
+            # is less than capacity_ratio. To save memory, we don't store the
+            # random numbers. Instead, we save the random number generator state.
+            rng_state = torch.get_rng_state()
+            for param_group in self.param_groups:
+                for p in param_group['params']:
+                    state = self.state[p]
+                    s = {}
+                    if 'rng_state' not in state:
+                        rng_state = torch.get_rng_state()
+                        s['rng_state'] = rng_state
+                    else:
+                        rng_state = state['rng_state']
+                        torch.set_rng_state(rng_state)
+                    n = p.numel()
+                    ratio = max(self._min_capacity / n, capacity_ratio)
+                    mask = torch.rand_like(p) >= ratio
+                    s['mask'] = mask
```

Comment:
It seems that mask generation can be put into apply_mask, so that we don't need to save all the masks in state.

Reply:
Changed as suggested